# <div style="text-align: center"> Applied Machine Learning </div>

###### <div style="text-align: right"> Udit Maniyar<br><br> ES16BTECH11024 </div>

## Theory

### 1) Kernels

<img src="images/1_a.jpg" style="width:800px;height:900px;">
<img src="images/1_b.jpg" style="width:800px;height:900px;">
<img src="images/1_cd.jpg" style="width:800px;height:900px;">
<img src="images/1_e.jpg" style="width:800px;height:900px;">

___

### 2) Perceptron Classifier

<img src="images/2_ab.jpg" style="width:600px;height:700px;">
<img src="images/2_c.jpg" style="width:600px;height:500px;">

___

### 3) Nueral Network

<img src="images/3_a.jpg" style="width:600px;height:700px;">
<img src="images/3a_pic.png" style="width:600px;height:400px;">

___

<img src="images/3_b.jpg" style="width:600px;height:400px;">

___

In [1]:
#Import All required Things
import numpy as np
import csv
import math
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
import timeit

## 4. SVMs

In [2]:
#Get Input from Postal Data
def postal_get_input():
    with open("data/postal.train") as f:
        training_data = f.readlines()
        #print(training_data)
        training_data = [line.rstrip("\n") for line in training_data]
        training_data = [list(map(float, line.split())) for line in training_data]
        training_data = np.array(training_data)
    train_set = []
    for i in range(len(training_data)):
        if training_data[i][0]==1 or training_data[i][0]==5:
            train_set.append(training_data[i])
            
            
    print ("Number of records in training data: %d" % len(train_set))
    
    test_set = []
    with open("data/postal.test") as f:
        test_data = f.readlines()
        test_data = [line.rstrip("\n") for line in test_data]
        test_data = [list(map(float, line.split())) for line in test_data]
        test_data = np.array(test_data)
        

    for i in range(len(test_data)):
        if test_data[i][0]==1 or test_data[i][0]==5:
            test_set.append(test_data[i])
    train_set = np.array(train_set)
    test_set = np.array(test_set)
    
    print ("Number of records in test data: %d" % len(test_set))


    return train_set,test_set

##### a) Linear Kernel

In [3]:
#Linear SVM
def linear_svm(training_data,test_data):
    x = training_data[:,1:]
    y = []
    for i in training_data:
        y.append(i[0])
    
    y = np.array(y)
    linear_svm = SVC(kernel = "linear")
    linear_svm.fit(x,y)
    
    
    train_outcome = linear_svm.predict(x)
    train_result = []
    for i in range(len(x)):
        if train_outcome[i] == y[i]:
            train_result.append(True)
        else:
            train_result.append(False)

    train_accuracy = train_result.count(True)/len(x)
    
    print("Training Accuracy: %.8f" % train_accuracy)
    train_error = 1-train_accuracy
    print("Training Error: %.8f" % train_error)
    
    outcome = linear_svm.predict(test_data[:,1:])
    result= []
    for i in range(len(test_data)):
        if outcome[i] == test_data[i][0]:
            result.append(True)
        else:
            result.append(False)
            
    accuracy = result.count(True)/len(test_data)
    print("Test accuracy: %.8f" % accuracy)
    test_error = 1-accuracy
    print("Test Error: %.8f" % test_error)

    support_vectors = linear_svm.support_vectors_
    print("The number of Support Vectors : ",len(support_vectors))

In [4]:
training_data, test_data = postal_get_input()

Number of records in training data: 1561
Number of records in test data: 424


In [5]:
print("Training Using Complete Data Set")
linear_svm(training_data,test_data)

Training Using Complete Data Set
Training Accuracy: 0.99615631
Training Error: 0.00384369
Test accuracy: 0.97877358
Test Error: 0.02122642
The number of Support Vectors :  28


##### (b) In continuation, train only using the first {50, 100, 200, 800} points with the linear kernel. Report the accuracy over the entire test set, and the number of support vectors in each of these cases.

In [6]:
data_50 = training_data[:50,:]
data_100 = training_data[:100,:]
data_200 = training_data[:200,:]
data_800 = training_data[:800,:]

print("----Training Using Only First 50 points----- ")
linear_svm(data_50,test_data)


print("----Training Using Only First 100 points----- ")
linear_svm(data_100,test_data)


print("----Training Using Only First 200 points----- ")
linear_svm(data_200,test_data)


print("----Training Using Only First 800 points----- ")
linear_svm(data_800,test_data)

----Training Using Only First 50 points----- 
Training Accuracy: 1.00000000
Training Error: 0.00000000
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  2
----Training Using Only First 100 points----- 
Training Accuracy: 1.00000000
Training Error: 0.00000000
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  4
----Training Using Only First 200 points----- 
Training Accuracy: 0.99500000
Training Error: 0.00500000
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  8
----Training Using Only First 800 points----- 
Training Accuracy: 0.99750000
Training Error: 0.00250000
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  14


The Accuracy for training on first 50,100,200,800 is just the same. But the number of support vectors is increseasing

###### (c) Polynomial Kernel

In [7]:
def polynomial_svm(training_data,test_data,c=1,deg = 2):
    x = training_data[:,1:]
    y = []
    for i in training_data:
        y.append(i[0])
    y = np.array(y)
    polynomial_svm = SVC(C = c, kernel = "poly",degree = deg,coef0 = 1, gamma = 1)
    polynomial_svm.fit(x,y)
   

    train_outcome = polynomial_svm.predict(x)
    train_result= []
    for i in range(len(x)):
        if train_outcome[i] == y[i]:
            train_result.append(True)
        else:
            train_result.append(False)

    train_accuracy = train_result.count(True)/len(x)
    print("Training Accuracy: %.8f" % train_accuracy)

    train_error = 1-train_accuracy
    print("Training Error: %.8f" % train_error)


    test_outcome = polynomial_svm.predict(test_data[:,1:])
    test_result= []
    for i in range(len(test_data)):
        if test_outcome[i] == test_data[i][0]:
            test_result.append(True)
        else:
            test_result.append(False)
            
    test_accuracy = test_result.count(True)/len(test_data)
    print("Test accuracy: %.8f" % test_accuracy)
    
    test_error = 1-test_accuracy
    print("Test Error: %.8f" % test_error)

    support_vectors = polynomial_svm.support_vectors_
    print("The number of Support Vectors : ",len(support_vectors))

In [8]:
polynomial_svm(training_data,test_data,c = 0.0001,deg = 2)

print("------------------")

polynomial_svm(training_data,test_data,c = 0.0001,deg = 5)

Training Accuracy: 0.99103139
Training Error: 0.00896861
Test accuracy: 0.98349057
Test Error: 0.01650943
The number of Support Vectors :  236
------------------
Training Accuracy: 0.99551570
Training Error: 0.00448430
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  26


In [9]:
polynomial_svm(training_data,test_data,0.001,2)

print("------------------")

polynomial_svm(training_data,test_data,0.001,5)

Training Accuracy: 0.99551570
Training Error: 0.00448430
Test accuracy: 0.98349057
Test Error: 0.01650943
The number of Support Vectors :  76
------------------
Training Accuracy: 0.99551570
Training Error: 0.00448430
Test accuracy: 0.97877358
Test Error: 0.02122642
The number of Support Vectors :  25


In [10]:
polynomial_svm(training_data,test_data,0.01,2)

print("------------------")

polynomial_svm(training_data,test_data,0.01,5)

Training Accuracy: 0.99551570
Training Error: 0.00448430
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  34
------------------
Training Accuracy: 0.99615631
Training Error: 0.00384369
Test accuracy: 0.97877358
Test Error: 0.02122642
The number of Support Vectors :  23


In [11]:
polynomial_svm(training_data,test_data,1,2)

print("------------------")

polynomial_svm(training_data,test_data,1,5)

Training Accuracy: 0.99679693
Training Error: 0.00320307
Test accuracy: 0.98113208
Test Error: 0.01886792
The number of Support Vectors :  24
------------------
Training Accuracy: 0.99679693
Training Error: 0.00320307
Test accuracy: 0.97877358
Test Error: 0.02122642
The number of Support Vectors :  21


###### (i) When C = 0.0001, training error is higher at Q = 5 : False <br>
###### (ii) When C = 0.001, the number of support vectors is lower at Q = 5. : True <br>
###### (iii) When C = 0.01, training error is higher at Q = 5. : False <br>
###### (iv) When C = 1, test error is lower at Q = 5. False <br>

#### d) Radial Basis Function

In [12]:
def radial_svm(training_data,test_data,c,gamm = "auto"):
    x = training_data[:,1:]
    y = []
    for i in training_data:
        y.append(i[0])
    y = np.array(y)
    polynomial_svm = SVC(C = c, kernel = "rbf",gamma=gamm)
    polynomial_svm.fit(x,y)
   

    train_outcome = polynomial_svm.predict(x)
    train_result= []
    for i in range(len(x)):
        if train_outcome[i] == y[i]:
            train_result.append(True)
        else:
            train_result.append(False)

    train_accuracy = train_result.count(True)/len(x)
    print("Training Accuracy: %.8f" % train_accuracy)

    train_error = 1-train_accuracy
    print("Training Error: %.8f" % train_error)


    test_outcome = polynomial_svm.predict(test_data[:,1:])
    test_result= []
    for i in range(len(test_data)):
        if test_outcome[i] == test_data[i][0]:
            test_result.append(True)
        else:
            test_result.append(False)

    test_accuracy = test_result.count(True)/len(test_data)

    print("Test accuracy: %.8f" % test_accuracy)
    test_error = 1- test_accuracy
    print("Test Error: %f" % test_error)
    support_vectors = polynomial_svm.support_vectors_
    print("The number of Support Vectors : ",len(support_vectors))

In [32]:
radial_svm(training_data,test_data,c = 0.01,gamm = 1)

print("------------------")

radial_svm(training_data,test_data,c = 1,gamm = 1)

print("------------------")

radial_svm(training_data,test_data,c = 100,gamm= 1)

print("------------------")

radial_svm(training_data,test_data,c = 10000,gamm = 1)

print("------------------")

radial_svm(training_data,test_data,c = 1000000,gamm = 1)

Training Accuracy: 0.99615631
Training Error: 0.00384369
Test accuracy: 0.97641509
Test Error: 0.023585
The number of Support Vectors :  403
------------------
Training Accuracy: 0.99551570
Training Error: 0.00448430
Test accuracy: 0.97877358
Test Error: 0.021226
The number of Support Vectors :  31
------------------
Training Accuracy: 0.99679693
Training Error: 0.00320307
Test accuracy: 0.98113208
Test Error: 0.018868
The number of Support Vectors :  22
------------------
Training Accuracy: 0.99743754
Training Error: 0.00256246
Test accuracy: 0.97641509
Test Error: 0.023585
The number of Support Vectors :  20
------------------
Training Accuracy: 0.99935939
Training Error: 0.00064061
Test accuracy: 0.97641509
Test Error: 0.023585
The number of Support Vectors :  17


###### Lowest Training Error is obtained when C = 1000000;Training Error = 0.00128123 <br>
###### Lowest Test Error is obtained when C = 100; Test Error is 0.018868 <br>

## Conclusion
A Support Vector Machine (SVM) is a discriminative classifier formally defined by a separating hyperplane. In other words, given labeled training data the algorithm outputs a hyperplane which maximized distance between two planes.
The data set consists of digits i.e class("1" = +1; "5" = -1) with 2 features/attributes.
The data set contains of 1561 training samples and 424 test data.
I have trained the data on Linear Kernel, Polynomial Kernels with different Degrre's, Gaussian Kernel with different values for C

- Linear Kernel

|  S.no  | No of Training Sample  | Training Error | Test Error |
|--------|------------------------|----------------|---------------|
| 1      | Full Training Set      | 0.00384369     | 0.02122642    |
| 2      | First 50 Samples Only  | 0.00000000     | 0.01886792    |
| 3      | First 100 Samples Only | 0.00000000     | 0.01886792    |
| 4      | First 200 Samples Only | 0.00500000     | 0.01886792    |
| 5      | First 800 Samples Only | 0.00250000     | 0.01886792    |

- Polynomial Kernel

|  S.no  | C      | Degree | Training Error | Test Error |
|--------|--------|--------|----------------|---------------|
| 1      | 0.0001 | 2      | 0.00896861     | 0.01650943    |
| 2      | 0.0001 | 5      | 0.00448430     | 0.01886792    |
| 3      | 0.001  | 2      | 0.00448430     | 0.01650943    |
| 4      | 0.001  | 5      | 0.00448430     | 0.02122642    |
| 5      | 0.01   | 2      | 0.00448430     | 0.01886792    |
| 6      | 0.01   | 5      | 0.00384369     | 0.02122642    |
| 7      | 1      | 2      | 0.00320307     | 0.01886792    |
| 8      | 1      | 5      | 0.00320307     | 0.02122642    |

- Gaussian Kernel

|  S.no  | C       | Training Error | Test Error |
|--------|---------|----------------|---------------|
| 1      | 0.01    | 0.00384369     | 0.021226      |
| 2      | 1       | 0.00448430     | 0.021226      |
| 3      | 100     | 0.00320307     | 0.018868      |
| 4      | 10000   | 0.00256246     | 0.018868      |
| 5      | 1000000 | 0.00128123     | 0.021226      |


___

### 5) GISETTE

In [14]:
#Get Gisette Data
def gisette_getdata():
    
    with open("data/gisette_train.data") as f:
        training_x = f.readlines()
        training_x = [line.rstrip("\n") for line in training_x]
        training_x = [list(map(float, line.split())) for line in training_x]
        training_x = np.array(training_x)
                    
    with open("data/gisette_train.labels") as f:
        training_y = f.readlines()
        training_y = [line.rstrip("\n") for line in training_y]
        training_y = [list(map(float, line.split())) for line in training_y]
        training_y = np.array(training_y)

    train_set = np.hstack((training_y,training_x))
    
    print ("Number of records in training data: %d" % len(train_set))
    
    
    with open("data/gisette_valid.data") as f:
        test_x = f.readlines()
        test_x = [line.rstrip("\n") for line in test_x]
        test_x = [list(map(float, line.split())) for line in test_x]
        test_x = np.array(test_x)

    with open("data/gisette_valid.labels") as f:
        test_y = f.readlines()
        test_y = [line.rstrip("\n") for line in test_y]
        test_y = [list(map(float, line.split())) for line in test_y]
        test_y = np.array(test_y)
    
    test_set = np.hstack((test_y,test_x))
    
    train_set = np.array(train_set)
    test_set = np.array(test_set)
    
    print ("Number of records in test data: %d" % len(test_set))

    return train_set,test_set

In [15]:
training_set,test_set  = gisette_getdata()

Number of records in training data: 6000
Number of records in test data: 1000


###### a) Linear SVM

In [16]:
linear_svm(training_set,test_set)

Training Accuracy: 1.00000000
Training Error: 0.00000000
Test accuracy: 0.97600000
Test Error: 0.02400000
The number of Support Vectors :  1084


###### b) RBF Kernel, Polynomial Kernel 

In [17]:
polynomial_svm(training_set,test_set,c = 1,deg = 2)

Training Accuracy: 1.00000000
Training Error: 0.00000000
Test accuracy: 0.97900000
Test Error: 0.02100000
The number of Support Vectors :  1755


In [18]:
radial_svm(training_set,test_set,c = 1,gamm = 1)

Training Accuracy: 1.00000000
Training Error: 0.00000000
Test accuracy: 0.50000000
Test Error: 0.500000
The number of Support Vectors :  6000


###### On the Gisette Data Set the all the three kernels i.e Linear Kernel, Polynomial Kernel and Gaussian Kernel Give Training Error of 0

## Conclusion
The problem is to separate the highly confusible digits ‘4’ and ‘9’.
The data set contains of 6000 training samples and 1000 samples of test data(Validation Set).
I have trained the data on Linear Kernel, Polynomial Kernel, Gaussian Kernel

|  S.no  | Type of Kernel                       | Training Error | Testing Error |
|--------|--------------------------------------|----------------|---------------|
| 1      | Linear Kernel                        | 0.00000000     | 0.02400000    |
| 2      | Polynomial Kernel(C = 1, Degree = 2) | 0.00000000     | 0.02100000    |
| 3      | Gaussian Kernel(Gamma = 0.001)       | 0.00000000     | 0.500000      |

___

## 6) Random Forests

(a) Write your own random forest classifier (this should be relatively easy, given you have written your own decision tree code) to apply to the Spam dataset [Data](https://web.stanford.edu/~hastie/ElemStatLearn//datasets/spam.data),[Information](https://web.stanford.edu/~hastie/ElemStatLearn//datasets/spam.info.txt). Use 30% of the provided data as test data and the remaining for training. Compare your results in terms of accuracy and time taken with Scikitlearn’s built-in random forest classifier.

### My Random Forest

In [27]:
#The following Decision Tree Code has been written in Assignment 1
class DecisionTree():
    tree = {}
    no_nodes = 0
    #Splits the data to two np arrays left and right based on the attribute and threshold
    def data_split(self,training_set,attr,thresh):
        left = []
        right = []
        
        for i in training_set:
            
            if i[attr] <= thresh:
                left.append(i)
            
            else:
                right.append(i)
        
        left = np.array(left)
        right = np.array(right)
        return left,right

    #Entropy calculation where left array contains l_one one's l_zero zero's and so on for right array
    def entropy_change(self, l_one,l_zero,r_one,r_zero):
        
        #Calculation for left array
        tot_left = l_one+l_zero
        left_part = 0
        
        if tot_left!=0:
        
            if l_one!=0:
                left_part += (l_one/tot_left)*math.log(l_one/tot_left)
            
            if l_zero!=0:
                left_part += (l_zero/tot_left)*math.log(l_zero/tot_left)


        #Calculation for left array
        tot_right = r_one + r_zero
        right_part = 0
        
        if tot_right!=0:
        
            if r_one!=0:
                right_part += (r_one/tot_right)*math.log(r_one/tot_right)
            
            if r_zero!=0:
                right_part += (r_zero/tot_right)*math.log(r_zero/tot_right)


                
        #Weighted Average
        total_length = tot_left + tot_right
        
        left_part = (tot_left/total_length) * left_part
        
        right_part = (tot_right/total_length) * right_part

        return -(left_part+right_part)
    
    #Ginni Index calculation where left array contains l_one one's l_zero zero's and so on for right array
    def calc_ginni(self, l_one,l_zero,r_one,r_zero):
        print("GINI")
        
        #Calculation for left array
        tot_left = l_one+l_zero
        left_part = 0
        
        if tot_left!=0:
        
            left_part += (l_one/tot_left)**2
            left_part += (l_zero/tot_left)**2
        
        left_part = 1-left_part

        
        #Calculation for left array
        tot_right = r_one + r_zero
        right_part = 0
        
        if tot_right!=0:
        
            right_part += (r_one/tot_right)**2
            right_part += (r_zero/tot_right)**2
        
        right_part = 1-right_part;

        
        #Weighted Average
        total = tot_left + tot_right
        
        left_part = (tot_left/total) * left_part
        
        right_part = (tot_right/total) * right_part

        return (left_part+right_part)
    
    
    #Selecting attributes based on entropy calculations
    def select_attribute(self,training_set,attributes,max_attributes,criterion = "entropy", min_entropy = 0):

        
        #Picking m attributes among all the attributes
        attributes_remaining = np.random.choice(len(attributes),max_attributes,replace = False)

        training_set = np.transpose(training_set)

        best_entropy = math.inf
        
        best_attr = []
        
        #For each Attribute in attributes remaining, pick the best attribute
        for i in attributes_remaining:
        
            attr_column = np.column_stack((training_set[i],training_set[-1]))
            attr_column = attr_column[np.argsort(attr_column[:, 0])]
            
            
            total_ones = sum(attr_column[:,-1])
            total_zeros = len(attr_column[:,-1])-total_ones

            if criterion == "ginni-index":
                ith_best = self.calc_ginni(total_ones,total_zeros,0,0)
                idx = math.inf
            else:
                ith_best  = self.entropy_change(total_ones,total_zeros,0,0)
                idx = math.inf

            
            ones_tillnow = 0
            zeros_tillnow = 0
            
            #In the ith attribute pick the best possible threshold
            for j in range(len(attr_column)):
                
                if attr_column[j][1]==0:
                    zeros_tillnow += 1
                
                else:
                    ones_tillnow +=1
                
                if criterion == "ginni-index":
                    current = self.calc_ginni(ones_tillnow, zeros_tillnow, total_ones-ones_tillnow,total_zeros-zeros_tillnow)
                else:
                    current  = self.entropy_change(ones_tillnow, zeros_tillnow, total_ones-ones_tillnow,total_zeros-zeros_tillnow)
                   
                
                if current < ith_best:
                
                    ith_best = current
                    idx = j

                    
            if ith_best < best_entropy:
                  
                    best_entropy = ith_best
                    
                    if idx!=math.inf:
                        best_attr = [i,attr_column[idx][0]]
                    
                    else:
                        best_attr = [i,idx]
            if best_entropy <= min_entropy:
                best_attr = [0,math.inf]

        return best_attr

    
    #Function call to build the tree and return the final tree in the form of a dicitonary 
    def build_tree(self,training_set,attributes_remaining,criterion,min_entropy,max_attributes):

        #Finding the best attribute and the corresponding threshold
        attr,thresh = self.select_attribute(training_set,attributes_remaining,max_attributes, criterion,min_entropy)
        
        #Split the data into two arrays left and right arrays based on the attr and threshold
        left,right = self.data_split(training_set,attr,thresh)

        #If the length of either left or right is 0 this means we cannot split it further we just return the most popular value
        if len(left)==0 or len(right)==0:
            
            nonzero = np.count_nonzero(training_set[:,-1])
            
            zeros = len(training_set[:,-1]) - nonzero

            if zeros >= nonzero:
                return 0.0
            else:
                return 1.0

        
        #Build Left Subtree and get the left subtree in left_tree
        left_tree = self.build_tree(left,attributes_remaining,criterion,min_entropy,max_attributes)
        
        #Build Right Subtree and get the right subtree in right_tree        
        right_tree = self.build_tree(right,attributes_remaining,criterion,min_entropy,max_attributes)

        
        #Store left subtree, right subtree, attribute selected at this node and threshold in a dictionary
        final = {"left":left_tree,"right": right_tree,"attr": attr,"thresh":thresh}
        self.no_nodes+=1
        return final


    #Calls buildtree function which returns the final tree 
    def learn(self, training_set,criterion,min_entropy,max_attributes):

        training_set = np.array(training_set)

        attributes_remaining = np.array(range(0,training_set.shape[1]-1))

        self.tree = self.build_tree(training_set,attributes_remaining,criterion,min_entropy,max_attributes)

    #Given a test instance we recursively go to left or right based on the attribute and threshold at the node
    def find_ans(self,node,test_instance):

        #Base Case
        if type(node)==float:
            return node

        #Recurse Down the tree
        if type(node)==dict:
            
            attr = node["attr"]
            thresh = node["thresh"]

            #Depending on the threhold value either go left or right
            if test_instance[attr] <=thresh:
                return self.find_ans(node["left"],test_instance)
            else:
                return self.find_ans(node["right"],test_instance)


    #Calls find ans which recursively finds the ans
    def classify(self, test_instance):
        
        result = self.find_ans(self.tree,test_instance)
        
        return result

In [28]:
#random_pick takes data and returns a data set with same length by picking by replacement from the original set
def random_pick(data):
    
    sample = np.random.choice(len(data),len(data),replace = True)
    
    present = np.zeros(len(data))
    
    for i in sample:
        present[i] += 1
    
    set1 = []
    for i in range(len(data)):
        if present[i] >= 1:
            for j in range(int(present[i])):
                set1.append(data[i])

    
    set1 = np.array(set1)

    
    return set1,present

In [29]:
#Get Input
def spam_get_input():
    with open("data/spam.data") as f:
        data = f.readlines()
        data = [line.rstrip("\n") for line in data]
        data = [list(map(float, line.split(" "))) for line in data]
        data = np.array(data)
        
    data = np.random.permutation(data)

    training_set = []
    test_set = []
    for i in range(len(data)):
        if i < round(0.7*len(data)):
            training_set.append(data[i])
        else:
            test_set.append(data[i])
                
    training_set = np.array(training_set)
    test_set = np.array(test_set)

    return training_set,test_set

In [30]:
#Function which runs the decision tree
def run_random_forest(no_trees, max_attributes, criterion,min_entropy = 0):

    training_set,test_set = spam_get_input()

    print("---------------------")

    print("No of Trees: ",no_trees,sep = " ")
    print("No of Maximum Attributes at Node Splitting: ",max_attributes,sep =" ")
    
    trees = []
    marker = []
    for j in range(no_trees):
        training_subset,mark = random_pick(training_set)
        marker.append(mark)

        tree = DecisionTree()
    
        # Construct a tree using training set
        tree.learn( training_subset, criterion, min_entropy,max_attributes)
        trees.append(tree)



    # Classify the test set using the trees we just constructed
    results = []
    for instance in test_set:
        result = []
        for i in range(no_trees):
            
            result.append(trees[i].classify(instance[:-1]))
    
        
        if result.count(True) > result.count(False):
            results.append( 1 == float(instance[-1]))
        else:
            results.append( 0 == float(instance[-1]))


    # Accuracy
    accuracy = float(results.count(True))/float(len(results))

    print("Test accuracy: %.8f" % accuracy)
    print("Test Error: %.8f" % (1-accuracy))
  

    #Calculate the OOB Error

    oob = []
    marker = np.array(marker)
    #print(marker.shape)
    for i in range(len(training_set)):
        result = []
        for j in range(no_trees):
            if marker[j][i] != 1:
                result.append(trees[j].classify(training_set[i][:-1]))
        
        if result.count(True) > result.count(False):
            oob.append(1==float(training_set[i][-1]))
        else:
            oob.append(0==float(training_set[i][-1]))
            
    oob_error =  1- float(oob.count(True))/float(len(oob))
    
    print("OOB Error: %.8f" % oob_error)
    print("Completed")
    print("---------------------")

### Random Forest Using Sklearn

In [31]:
def sklearn_Randomforest(no_trees,m):
    training_set,test_set = spam_get_input()
    
    print("---------------------")    
    print("No of Trees: ",no_trees,sep = " ")
    print("No of Maximum Attributes at Node Splitting: ",m,sep =" ")
    
 
    random_forest = RandomForestClassifier(n_estimators = no_trees,max_features= m, n_jobs =-1)
    training_set = np.array(training_set)
    x = training_set[:,:-1]
    y = training_set[:,-1]
    random_forest.fit(x,y)
    #print(random_forest.feature_importances_)

    predicted_result = random_forest.predict(test_set[:,:-1])
    actual_result = test_set[:,-1]
    
    # Classify the test set using the Random forest we
    results = []
    for i in range(len(test_set)):
        results.append( predicted_result[i] == actual_result[i])
    

    # Accuracy
    accuracy = float(results.count(True))/float(len(results))

    print("accuracy: %.8f" % accuracy)
    print("Completed")
    print("---------------------")

#### Comparing My Random Forest With Sklearns Random Forest

In [24]:
#Running My random Forest 
start = timeit.default_timer()

run_random_forest(no_trees=25,max_attributes =8, criterion = "entropy")

stop = timeit.default_timer()
print('Time: ', stop - start)  

start = timeit.default_timer()

run_random_forest(no_trees=25,max_attributes =22, criterion = "entropy")

stop = timeit.default_timer()
print('Time: ', stop - start)  

---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  8
Test accuracy: 0.93623188
Test Error: 0.06376812
OOB Error: 0.04036014
Completed
---------------------
Time:  45.40841455300688
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  22
Test accuracy: 0.94637681
Test Error: 0.05362319
OOB Error: 0.02732071
Completed
---------------------
Time:  116.17177796500619


In [25]:
#Running Sklearn's Random Forest

start = timeit.default_timer()

sklearn_Randomforest(no_trees = 25,m = 8)

stop = timeit.default_timer()
print('Time: ', stop - start)  

start = timeit.default_timer()

sklearn_Randomforest(no_trees = 25,m = 22)

stop = timeit.default_timer()
print('Time: ', stop - start)  

---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  8
accuracy: 0.94782609
Completed
---------------------
Time:  0.29053281500091543
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  22
accuracy: 0.95000000
Completed
---------------------
Time:  0.40412301600008504


Conclusion
----------
- Sklearns Random Forest takes very less time compared to my implementation of Random Forest.
- Time taken by my random forest when m = 8 and no of trees = 25 is on average 45 secs whereas the time taken by sklearns Random Forest for the same input is always less than 1 sec ~0.3
- Time taken by my random forest when m = 22 and no of trees = 25 is on average 115 secs whereas the time taken by sklearns Random Forest for the same input is less than 1sec ~0.35
- sklearns random Forest trains the trees paralelly when n_jobs is set to -1

- Accuracy obtained by my random Forest for m = 8 and no of trees = 25 is on average 94 and the accuracy obtained by sklearn's Random Forest on average is 95%
- Accuracy given by both is almost the same between 92 - 96% for both. 

### (b) Explore the sensitivity of Random Forests to the parameter m (no features used for best split).

In [38]:
# We try to run our random forest for different values of m. m  = (2, 6, 8, .... 50, 54)
for j in range(2,57,4):
    print("------",j,"------")
    run_random_forest(no_trees=25,max_attributes =j, criterion = "entropy")

------ 2 ------
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  2
Test accuracy: 0.92101449
Test Error: 0.07898551
OOB Error: 0.05371003
Completed
---------------------
------ 6 ------
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  6
Test accuracy: 0.94637681
Test Error: 0.05362319
OOB Error: 0.04160199
Completed
---------------------
------ 10 ------
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  10
Test accuracy: 0.93913043
Test Error: 0.06086957
OOB Error: 0.03477181
Completed
---------------------
------ 14 ------
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  14
Test accuracy: 0.94492754
Test Error: 0.05507246
OOB Error: 0.03290903
Completed
---------------------
------ 18 ------
---------------------
No of Trees:  25
No of Maximum Attributes at Node Splitting:  18
Test accuracy: 0.95072464
Test Error: 0.04927536
OOB Error: 0.03352996
C

The test Error is more in the beginning when the m is sqrt(no of attributes) the test error is low
again in between 20-30 we see a slight peek and then from there the test error almost remains the same. For the exact plot of training error vs m see the plot in the following question.

### c) Plot the OOB (out-of-bag) error and the test error against a suitably chosen range of values for m.

#### Out of Bag Estimate

After creating the classifiers (p trees), for each (xi,yi) in the original training set i.e. T, select all Tk which does not include (xi,yi). FOr each point (xi,yi) in select all the trees which were build without using (xi,yi) and predict (xi,yi) on only those trees. DO the above step for all the data points in T

Out-of-bag estimate for the generalization error is the error rate of the out-of-bag classifier on the training set (compare it with known yi's).

The values for the below graph have been generated by taking no of trees  = 25.
<img src="images/6c.png" style="width:400px;height:300px;">